Internal: Migrating behat to headless chrome

bojanbogdanovic opened the pull request:
## Problem
Our behat tests are inconsistent, slow and outdated; due the fact that we use an old firefox webdriver (2.48.2) in combination with selenium. The modern browsers have updated or even introduced a complete new JavaScript and rendering engine, this introduces inconsistencies in expecations for the tests. To guarantee the quality of our distribution, we need to start testing against modern browsers.

## Solution
By using headless chrome: our behat tests will be more performant and leverage a recent chrome version.
We choose for the "zenika/alpine-chrome" container, because it's well maintained, popular and well documented, also it's being used by known local development tools.

## Issue tracker
N.A

## Theme issue tracker
N.A

## How to test
- [ ] Use [goalgorilla/drupal_social:feature/headless-chrome](https://github.com/goalgorilla/drupal_social/pull/375)
- [ ] Use [goalgorilla/open_social_scripts:feature/headless-chrome](https://github.com/goalgorilla/open_social_scripts/pull/45)
- [ ] Use [goalgorilla/open_social_dev:feature/headless-chrome](https://github.com/goalgorilla/open_social_dev/pull/15)
- [ ] Run `composer install`
- [ ] Run `docker-compose up -d`
- [ ] Run `docker exec -it social_web bash /var/www/scripts/social/install/install_script.sh --with-optional --localsettings`
- [ ] Run `docker exec -it social_web sh /var/www/scripts/social/behatstability.sh DS-233`
- [ ] Failed tests will generate screenshot and html files in the dir: `/var/www/html/profiles/contrib/social/tests/behat/logs`

See: https://github.com/goalgorilla/open_social/blob/cfaa9668e1d0ad88c54d2b84b24f540116de495d/tests/behat/README.md

## Definition of done
### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Screenshots
N.A

## Release notes
N.A

## Change Record
N.A

## Translations
N.A

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

tbsiqueira commented:
Great contribution!

This will allow us to fix the commented code at this test. https://github.com/goalgorilla/open_social/pull/2830/files#diff-babfa4388fb1d4e23ecfcb0434f215d7154a476fbb461a5ae89b49cf4f5cd33eR38

It was necessary because of the old firefox driver

tbsiqueira commented:
We are currently introducing a lot of nice structural changes to our test environment, and I would love to see some more documentation over why we did it and how can it affect our community users. 

We should open a meta ticket on our issue queue to let the community aware of the changes that we are introducing since they might affect them, not to mention the nice new improvements ( like the pictures/html folder ).

It will be nice to review and update the following d.org pages with our new standards.

https://www.drupal.org/docs/drupal-distributions/open-social/contribute-to-open-social
https://www.drupal.org/docs/distributions/open-social/how-to-run-the-behat-tests-locally

Great job people!

tbsiqueira reviewed (changes requested):
Please check my comments about the necessary documentation that this change requires.

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

bojanbogdanovic commented:
I have added '@javascript' tag on some feature tests, because it was giving undefined jQuery error. The javascript tag enables JS on the associated feature/scenario.
Some tests were failing because it was trying to fill form elements that were hidden due collapsible details, e.g. the following step resolves the associated errors: `I click the xth "0" element with the css "#attachments summary"`.
Headless chrome is faster, so I removed some redundant delayed steps e.g: `I wait for "3" seconds`.

We still have two failing tests:

- `stability-1&&~DS-1136&&~DS-3605` This seems a driver error, I could not reproduce this locally with docker. 
- `notifications`

tbsiqueira commented:
DO NOT MERGE IT YET

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

navneet0693 commented on .github/workflows/behat-stability-1-upgrade-path.yml:
```diff
@@ -83,4 +83,4 @@ jobs:
       # Run the upgrade path behat tests
       - name: Run Integration tests
-        run: docker exec -i social_ci_behat sh /var/www/scripts/social/behatstability.sh "stability-1&&~DS-1136&&~DS-3605 --stop-on-failure --strict  --colors"
+        run: docker exec -i social_ci_web sh /var/www/scripts/social/behatstability.sh "stability-1&&~DS-1136&&~DS-3605 --stop-on-failure --strict  --colors"
```
All of the above steps are added except VNC part. The new branch of drupal_social being used is https://github.com/goalgorilla/drupal_social/tree/feature/upgrade-path-headless-chrome

navneet0693 replied:
@tbsiqueira Please repeat the steps of upgrade path tests on your local and see if are able to spot the error remaining.

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

navneet0693 commented on .github/workflows/behat-stability-1-upgrade-path.yml:
```diff
@@ -79,6 +79,11 @@ jobs:
       - name: Run the updates
         run: docker exec -i social_ci_web bash /var/www/scripts/social/install/update.sh
 
+      - name: Restart Docker containers
```
@Ressinel Why do we need to restart the containers?

navneet0693 replied:
@Ressinel See the output of restarting container here: https://github.com/goalgorilla/open_social/runs/7558047207?check_suite_focus=true#step:14:1 you will see that the containers are getting re-created.

Ressinel replied:
@navneet0693 Actually, I am not completely sure why it doesn't work without restarting containers. But it is only one decision when the upgrade path works with reproducible error.

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5

open-social-tugboat commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3010-oepljhp7e8tn8i7fnpgixv1ndd44yjzl.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/62b454976630a8a12ad6a2b5
